core/src/replay_stage.rs

```diff
@@ -708,8 +687,28 @@ impl ReplayStage {
             );
         }
         inc_new_counter_info!("replay_stage-tower_duration", ms as usize);
+        rv
+    }
 
-        votable
+    #[allow(clippy::type_complexity)]
+    fn check_fork(
+        ancestors: &HashMap<u64, HashSet<u64>>,
+        tower: &Tower,
+        best_bank: Option<&(u128, Arc<Bank>, HashMap<u64, StakeLockout>, u64)>,
+    ) -> Option<(Arc<Bank>, u64)> {
+        let (_, bank, stake_lockouts, total_staked) = best_bank?;
+        let vote_threshold =
+            tower.check_vote_stake_threshold(bank.slot(), &stake_lockouts, *total_staked);
+        if !vote_threshold {
+            inc_new_counter_info!("replay_stage-fork_selection-heavy_bank_threshold", 1);
+            return None;
+        }
+        let is_locked_out = tower.is_locked_out(bank.slot(), &ancestors);
+        if is_locked_out {
+            inc_new_counter_info!("replay_stage-fork_selection-heavy_bank_lockout", 1);
+            return None;
```

Comment:
What happens if the heaviest fork dies and stops making process before your lockout expires?

Reply:
@carllin how do we know its "expired"?  The way this is written is that it will vote on it, and then it will be filtered out as an already voted fork.

Reply:
From the validiator's perspective, if they see the heaviest bank on some fork here, but the validator is locked out on another fork, then it returns None here. if that heaviest fork then stops making progress, we will keep returning None here and never vote?

Reply:
@carllin how would it stop making progress?  It's a frozen bank we haven't voted on.

Reply:
You don't need to vote on the frozen heaviest bank in order to be locked out on it. If you've voted on any other bank > frozen heaviest bank, you will be locked out on the heaviest bank. This will stop progress unless the heaviest bank changes (which doesn't seem is guaranteed to happen)

Reply:
@carllin ah right, you can't vote backwards.  This is why I thought we might need to consider leafs.

Reply:
@carllin I added a filter to only consider leafs of banks

Reply:
fixed, only vote on banks that are newer that the last vote